src/mr/coordinator.go

```diff
@@ -1,33 +1,131 @@
+	"net"
+	"net/http"
+	"net/rpc"
+	"os"
+	"sync"
+)
+
+// Coordinator can be initialized with a list of files, and n, the number of reduce tasks
+// That is to say, there are unlimited number of map task, as long as there are worker available.
+// However, each worker only supposed to generate n output files. mr-X-0, mr-X-1, ... mr-X-(n-1)
+//
+// 1) The coordinator will hold a channel of all the input files.
+// When there is a worker available, the worker will ask coordinator to assign a map task (a file) to it.
+// 2) The coordinator will also need to record all the in-flight tasks and check the health of the workers periodically.
+// If a worker died, the task will be added back to the channel, so that another healthy worker can pick it up.
+//
+//	 Q: If a worker, say worker 5, successfully finished a task, and generate out mr-5-0,
+//		but died while generating mr-5-100, how to keep mr-5-0 but delete the partially finished mr-5-100?
+//
+//		A: We will rename the worker, next time it get named as worker 6, so, ideally, when coordinator detects worker 6 died,
+//		before coordinator reassign worker 6's input file, coordinator will clean up worker 6's staled output files. But in this lab,
+//		we don't need to. Coordinator can have a field to record all the finished intermediate files.
+//		The reduce job is only going to work on these intermediate files.
 type Coordinator struct {
 	// Your definitions here.
+	// Mutex lock that avoid data racing
+	// TODO(ywmei): When to use mutex, when to use sync.Map?
+	mu sync.Mutex
+	// nReduce is the number of reduce jobs (col numbers)
+	nReduce int
+	// nMap is the number of map jobs (row numbers)
+	nMap int
+	// mr is the mapping between input file names to the num of worker do the map job.
+	mr map[string]int
+	// A channel of all the files that need to be mapped
+	files chan string
+	// A hashmap of all the in-flight map jobs. key is the map-worker idx (int), value is the file name (string) that the worker is working on.
+	inFlightMapJobs map[string]bool
+	// A hashmap of finished map jobs, where key is the file name string that has been finished, value is the map-worker idx (int).
+	finishedMapJobs map[string]bool
+	// A hashmap of all the in-flight reduce jobs. key is the reduce-worker idx (int), value is the file name (string) that the worker is working on.
+	inFlightReduceJobs map[string]bool
+	// A hashmap of finished reduced jobs, where key is the file name string that has been finished, value is the reduce-worker idx (int).
+	finishedReduceJobs map[string]bool
+}
 
+// MakeCoordinator create a Coordinator.
+// main/mrcoordinator.go calls this function.
+// nReduce is the number of reduce tasks to use.
+func MakeCoordinator(files []string, nReduce int) *Coordinator {
+	c := Coordinator{
+		nReduce:            nReduce,
+		nMap:               len(files),
+		mr:                 map[string]int{},
+		files:              make(chan string, len(files)),
+		inFlightMapJobs:    make(map[string]bool),
```

Comment:
We might need to have a lock (or a few locks for coordinator). 

One way is to use a read mutex lock, which is optimized in reading. But the question is do we need to lock the entire Coordinator struct?

The other way is to use, say, 4 different syncMap, if you read the doc, syncMap is optimized in small amount of writing, so multiply workers can write to the map together, but when the map need to expand, it locks all the writer/readers.

Or, do we really need a lock? If instead of having a coordinator for both map and reduce, maybe we can have one pusher, which job is only to fan out the map job, and one reciever, which job is only to fan in the reduce job.

Overall, the only way to test out which design is better is to run some benchmark test, collect some data and plot the metrics would be interesting.